frappe/rate_limiter.py

```diff
@@ -82,37 +82,50 @@ def respond(self):
 		if self.rejected:
 			return Response(_("Too Many Requests"), status=429)
 
-def rate_limit(key: str, limit: Union[int, Callable] = 5, seconds: int= 24*60*60, methods: Union[str, list]='ALL'):
+def rate_limit(key: str = None, limit: Union[int, Callable] = 5, seconds: int = 24*60*60, methods: Union[str, list] = 'ALL', ip_based: bool = True):
 	"""Decorator to rate limit an endpoint.
 
 	This will limit Number of requests per endpoint to `limit` within `seconds`.
 	Uses redis cache to track request counts.
 
-	:param key: Key is used to identify the requests uniqueness
+	:param key: Key is used to identify the requests uniqueness (Optional)
 	:param limit: Maximum number of requests to allow with in window time
 	:type limit: Callable or Integer
 	:param seconds: window time to allow requests
 	:param methods: Limit the validation for these methods.
 		`ALL` is a wildcard that applies rate limit on all methods.
 	:type methods: string or list or tuple
+	:param ip_based: flag to allow ip based rate-limiting
+	:type ip_based: Boolean
 
 	:returns: a decorator function that limit the number of requests per endpoint
 	"""
 	def ratelimit_decorator(fun):
 		@wraps(fun)
 		def wrapper(*args, **kwargs):
 			# Do not apply rate limits if method is not opted to check
-			if methods != 'ALL' and frappe.request.method.upper() not in methods:
+			if methods != 'ALL' and frappe.request and frappe.request.method and frappe.request.method.upper() not in methods:
 				return frappe.call(fun, **frappe.form_dict or kwargs)
 
 			_limit = limit() if callable(limit) else limit
 
-			identity = frappe.form_dict[key]
+			ip = frappe.local.request_ip
+
+			if key is None and ip_based is False:
+				frappe.throw(_('Either key or IP flag is required.'))
+			elif key is None:
+				identity = ip
+			elif ip_based is False:
+				identity = frappe.form_dict[key]
+			else:
+				user_key=frappe.form_dict[key]
+				identity = ":".join([ip, user_key])
+
```

Comment:
Order does not seem correct. You may need to use use combination of ip and user_key, if both exists. But this does not work that way..

if ip and user_key:
    identity = ":".join([ip, user_key])
identity = identity or ip or user_key

if not identity:
    frappe.throw(_('Either key or IP flag is required.'))

Reply:
This approach will fail since ip will never be empty.
`ip = frappe.local.request_ip`

Reply:
My bad,  I have used ip instead of `ip_based`.. changed that now

But order needs a change, isn't it?

Reply:
Yeah Thanks for the suggestion. Cleaned up a bit.